nova/cmd/manage.py

```diff
@@ -3183,13 +3184,183 @@ def refresh(self, instance_uuid=None, volume_id=None, connector_path=None):
             return 1
 
 
+class SAPCommands(object):
+
+    @action_description(
+        _("Detects changes on the flavors extra_specs and syncs the flavor of"
+          "instances and request_spec with the last extra_specs. It skips "
+          "instances that are resizing."))
+    @args('--max-count', metavar='<max_count>', dest='max_count',
+          help='Maximum number of instances to process. If not specified, all '
+               'instances in each cell will be mapped in batches of 50. '
+               'If you have a large number of instances, consider specifying '
+               'a custom value and run the command until it exits with '
+               '0 or 4.')
+    @args('--verbose', action='store_true', dest='verbose', default=False,
+          help='Provide verbose output during execution.')
+    @args('--dry-run', action='store_true', dest='dry_run', default=False,
+          help='Runs the command and prints output but does not commit any '
+               'changes. The return code should be 4.')
+    @args('--instance', metavar='<instance_uuid>', dest='instance_uuid',
+          help='UUID of a specific instance to process. If specified '
+               '--max-count has no effect.')
+    def sync_instances_flavor(self, max_count=None, verbose=False,
+                              dry_run=False, instance_uuid=None):
+        output = lambda msg: None
+        if verbose:
+            output = lambda msg: print(msg)
+
+        ctxt = context.get_admin_context()
+        flavors = objects.FlavorList.get_all(ctxt)
+
+        if instance_uuid:
+            max_count = 1
+            unlimited = False
+        elif max_count is not None:
+            try:
+                max_count = int(max_count)
+            except ValueError:
+                max_count = -1
+            unlimited = False
+            if max_count < 1:
+                print(_('Must supply a positive integer for --max-count.'))
+                return 127
+        else:
+            max_count = 50
+            unlimited = True
+            output(_('Running batches of %i until complete') % max_count)
+
+        def _find_flavor(instance):
+            for flavor in flavors:
+                if instance.flavor and instance.flavor.name == flavor.name:
+                    return flavor
+            return None
+
+        def _should_update(instance, flavor):
+            # exclude resizing instances
+            if (instance.task_state and 'resize_' in instance.task_state) \
+                    or instance.vm_state == vm_states.RESIZED:
+                return False
```

Comment:
I think we should exclude all instances currently getting worked on i.e. every instance having a non-empty `task_state`. This would then also exclude instances currently in build-up or getting live-migrated. What do you think?

Reply:
I think I'd still keep the `instance.vm_state == vm_states.RESIZED` in there, mainly because I don't feel comfortable with playing with the flavors when there is a new and an old one that could both be used still.